src/gui/guiiconprovider.cpp

```diff
@@ -135,13 +148,87 @@ QString GuiIconProvider::getIconPath(const QString &iconId)
         return path;
     }
 #endif
-    return IconProvider::getIconPath(iconId);
+    return IconProvider::getIconPath(iconId, overrideTheme);
 }
 
-
 void GuiIconProvider::configure()
 {
 #if (defined(Q_OS_UNIX) && !defined(Q_OS_MAC))
     m_useSystemTheme = Preferences::instance()->useSystemIconTheme();
 #endif
 }
+
+int GuiIconProvider::getThemeFlags()
+{
+    // QPalette::Base is used for the background of the Treeview
+    QPalette pal = QApplication::palette();
+    QColor color = pal.color(QPalette::Active, QPalette::Base);
+    bool darkTheme = (color.lightness() < 127);
+    int themeFlags = darkTheme ? DarkTheme : LightTheme;
+
+    const Preferences *const pref = Preferences::instance();
+    themeFlags |= pref->useLargeStatusIcons() ? LargeStatusIcons : SmallStatusIcons;
+
+    return themeFlags;
+}
+
+QIcon GuiIconProvider::getStatusIcon(const QString &iconId)
+{
+    return getStatusIcon(iconId, BitTorrent::TorrentState::Unknown, true);
+}
+
+QIcon GuiIconProvider::getStatusIcon(const QString &iconId, BitTorrent::TorrentState state)
+{
+    return getStatusIcon(iconId, state, false);
+}
+
+QIcon GuiIconProvider::getStatusIcon(const QString &iconId, BitTorrent::TorrentState state, bool ignoreState)
+{
+    static int cachedIconFlags = 0;
+    static QHash<QPair<QString, BitTorrent::TorrentState>, QIcon> cache;
+    int theme = getThemeFlags();
```

Comment:
Because that's how the dark theme was detected previously, and I simply extended it. Didn't know about `QEvent::PaletteChange` - I tried to find some event that would indicate a theme change but didn't know what to look for.

Now I'll see about hooking into it.